incubator/hnc/pkg/stats/stats.go

```diff
@@ -0,0 +1,187 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type counter int32
+
+type object struct {
+	gk              schema.GroupKind
+	totalReconciles counter
+	curReconciles   counter
+	apiWrites       counter
+}
+
+type objects map[schema.GroupKind]*object
+
+type stat struct {
+	// actionID is the number of controller actions devided by idles
+	actionID counter
+
+	// totalHierConfigReconciles is the total number of HierarchyConfig reconciliations.
+	totalHierConfigReconciles counter
+
+	// curHierConfigReconciles is the currently undergoing number of HierarchyConfig reconciliations.
+	curHierConfigReconciles counter
+
+	// hierConfigWrites is the total number of HierarchyConfig writes.
+	hierConfigWrites counter
+
+	// namespaceWrites is the total number of Namespace writes.
+	namespaceWrites counter
+
+	objects objects
+}
+
+var stats stat
+
+// StartHierConfigReconcile updates stats when hierarchyConfig
+// reconciliation starts.
+func StartHierConfigReconcile() {
+	stats.totalHierConfigReconciles.incr()
+	stats.curHierConfigReconciles.incr()
+}
+
+// StopHierConfigReconcile updates stats when hierarchyConfig
+// reconciliation finishes.
+func StopHierConfigReconcile() {
+	stats.curHierConfigReconciles.decr()
+}
+
+// StartObjReconcile updates the stats for objects with common GK
+// when an object reconciliation starts.
+func StartObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
+	if _, ok := stats.objects[gk]; !ok {
+		stats.objects[gk] = &object{
+			gk: gk,
+		}
+	}
+	stats.objects[gk].totalReconciles.incr()
+	stats.objects[gk].curReconciles.incr()
+}
+
+// StopObjReconcile updates the stats for objects with common GK
+// when an object reconciliation finishes.
+func StopObjReconcile(gk schema.GroupKind) {
+	if stats.objects == nil {
+		stats.objects = make(map[schema.GroupKind]*object)
+	}
+	if _, ok := stats.objects[gk]; !ok {
+		stats.objects[gk] = &object{
+			gk: gk,
+		}
+	}
+	stats.objects[gk].curReconciles.decr()
+}
+
+// WriteNamespace updates stats when writing namespace instance.
+func WriteNamespace() {
+	stats.namespaceWrites.incr()
+}
+
+// WriteHierConfig updates stats when writing hierarchyConfig instance.
+func WriteHierConfig() {
+	stats.hierConfigWrites.incr()
+}
+
+// WriteObject updates the object stats by GK when writing the object.
+func WriteObject(gk schema.GroupKind) {
+	stats.objects[gk].apiWrites.incr()
+}
+
+func initStats() {
+	objects := make(map[schema.GroupKind]*object)
+	for _, gvk := range config.GVKs {
+		objects[gvk.GroupKind()] = &object{
+			gk: gvk.GroupKind(),
+		}
+	}
+	stats = stat{
+		actionID: 1,
+		objects:  objects,
+	}
+}
+
+// StartLoggingActivity generates logs for performance testing.
+func StartLoggingActivity() {
+	initStats()
+	log := ctrl.Log.WithName("reconcileCounter")
+	var total, lastTotal, lastCur counter = 0, 0, 0
+	working := false
+	go func() {
+		// run forever
+		for {
```

Comment:
Suggestion: factor this out into its own function, then you can get rid of two extra levels of indentation.